lib/src/main/java/com/auth0/jwt/JWTCreator.java

```diff
@@ -164,6 +176,78 @@ public Builder withJWTId(String jwtId) {
             return this;
         }
 
+        /**
+         * Add a Map of Claim values. Only primitives and collections are supported as
+         * values in the map. Warning: Custom classes are not allowed.
+         *
+         * @param name
+         * @param values
+         * @throws IllegalArgumentException
+         */
+        public Builder withClaim(String name, Map<String, Object> values) throws IllegalArgumentException {
+            assertNonNull(name);
+            if (!validateClaim(values)) {
+                throw new IllegalArgumentException(
+                        "Unsupported claim type found in passed collection. Refer supported types.");
+            }
+            addClaim(name, values);
+            return this;
+        }
+
+        /**
+         * Add a Collection of Claim values. Only primitives and supported array types
+         * are supported as values in the Collection. Warning: Custom classes are not
+         * allowed.
+         *
+         * @param name
+         * @param values
+         * @throws IllegalArgumentException
+         */
+        public Builder withClaim(String name, Collection<Object> values) throws IllegalArgumentException {
+            assertNonNull(name);
+            if (!validateClaim(values)) {
+                throw new IllegalArgumentException(
+                        "Unsupported claim type found in passed collection. Refer supported types.");
+            }
+            addClaim(name, values);
+            return this;
+        }
+
+        private static boolean validateClaim(Map<String, Object> claim) {
+            if (claim == null || claim.isEmpty()) {
+                return false;
+            }
+            Collection<Object> values = claim.values();
+            return validateClaim(values);
+        }
+
+        private static boolean validateClaim(Collection<Object> values) {
+            if (values == null || values.isEmpty()) {
+                return false;
+            }
+            for (Object value : values) {
+                if (!isValidClaimValue(value)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        /**
+         * We are allowing value can be null at this point which will be handled at
+         * {@link #addClaim(String, Object)}
+         */
+        private static boolean isValidClaimValue(Object value) {
+            if (value == null) {
+                return true;
+            }
+            Class<?> clazz = value.getClass();
+            if (clazz.isArray() && SUPPORTED_ARRAY_TYPES.contains(clazz)) {
+                return true;
+            }
+            return SUPPORTED_TYPES.contains(clazz);
+        }
```

Comment:
I suggest, instead of returning boolean, throw IllegalArgumentException right from here.
The benefit is that you can provide the class and value of the offending claim right in the exception message, which GREATLY simplifies the troubleshooting.

Reply:
Good one, I've updated the PR